src/rule-applier.js

```diff
@@ -124,10 +151,86 @@ module.exports = class RuleApplier {
     const getRelatedCardPromise = this.issueDataProvider.getRelatedCardInKanban(issueId);
     const relatedCard = await getRelatedCardPromise;
 
-    this.githubApiClient.projects.moveProjectCard({
+    this.githubApiClient.projects.moveCard({
       card_id: relatedCard.id,
       position: 'bottom',
       column_id: this.config.kanbanColumns.doneColumnId
     });
   }
+
+  /**
+   * @param {Context} context
+   */
+  async applyRuleD1 (context) {
+    const issueId = context.payload.issue.number;
+    const newLabel = context.payload.label;
+
+    for (const label of context.payload.issue.labels) {
+      if (this.isAutomaticLabel(label) && newLabel.id !== label.id) {
+        this.logger.info(`[Rule Applier] D1 - Remove label ${label.name}`);
+
+        await this.githubApiClient.issues.removeLabel({
+          issue_number: issueId,
+          owner: context.payload.repository.owner.login,
+          repo: context.payload.repository.name,
+          name: label.name
+        })
+      }
+    }
+  }
+
+  /**
+   * @param {Context} context
+   */
+  async applyRuleD2 (context) {
+    const issueId = context.payload.issue.number;
+    this.logger.info(`[Rule Applier] D2 - Add label ${this.config.labels.todo.name}`);
+
+    await this.githubApiClient.issues.addLabels({
+      issue_number: issueId,
+      owner: context.payload.repository.owner.login,
+      repo: context.payload.repository.name,
+      labels: { labels: [this.config.labels.todo.name] }
+    })
+  }
+
+  /**
+   * @param {Context} context
+   */
+  async applyRuleD3 (context) {
+    const issueId = context.payload.issue.number;
+
+    for (const label of context.payload.issue.labels) {
+      if (this.isAutomaticLabel(label)) {
+        this.logger.info(`[Rule Applier] D3 - Remove label ${label.name}`);
+
+        await this.githubApiClient.issues.removeLabel({
+          issue_number: issueId,
+          owner: context.payload.repository.owner.login,
+          repo: context.payload.repository.name,
+          name: label.name
+        })
+      }
+    }
+  }
+
+  /**
+   * @param {Context} context
+   */
+  async applyRuleD4 (context) {
+    const issueId = context.payload.issue.number;
+
+    for (const label of context.payload.issue.labels) {
+      if (this.isAutomaticLabel(label)) {
+        this.logger.info(`[Rule Applier] D3 - Remove label ${label.name}`);
+
+        await this.githubApiClient.issues.update({
+          issue_number: issueId,
+          owner: context.payload.repository.owner.login,
+          repo: context.payload.repository.name,
+          state: 'open'
+        })
+      }
+    }
+  }
```

Comment:
This looks so duplicating, wdyt @PierreRambaud ?

Reply:
Yep :+1:

Reply:
Then would it be possible to try to do something to avoid duplicate code ? Sometimes  it's too overkill but there it looks like not